### PR TITLE
Fix ChangeText not being fully applied for trigger effects + traits gained by static

### PR DESCRIPTION
**Pull request by tool4ever**
E. g. _Artificial Evolution_ on _Perimeter Sergeant_

### CODE REVIEW

**Comment by Hanmac** [comment]
LockTokenScript$ isn't needed for such Keywords because they should NOT be affected by Change Text

**Comment by tool4ever** [comment]
> LockTokenScript$ isn't needed for such Keywords because they should NOT be affected by Change Text
> 
> That us why it's important what Part of the Keyword can be affected by Text Changes and what Parts not

yea the KW won't be modified
but in TriggerHandler it doesn't know that the effect belongs to one
should I try to add some skip logic there?

**Comment by Hanmac** [comment]
the thing is that the TriggerHandler should not need to care

the Card Traits are updated there:

```java

    public void updateChangedText() {
        final List<CardTraitBase> allAbs = ImmutableList.<CardTraitBase>builder()
            .addAll(manaAbilities)
            .addAll(nonManaAbilities)
            .addAll(triggers)
            .addAll(replacementEffects)
            .addAll(staticAbilities)
            .build();
        for (final CardTraitBase ctb : allAbs) {
            if (ctb.isIntrinsic()) {
                ctb.changeText();
            }
        }
    }
```

noticed how it doesn't touch keywords?

because it is more clean to substitute the Keyword directly like from "Protection:Red" to "Protection:Blue" than to and rebuilt the Keyword than to try to update the Keyword parts (where at this point you might not know which parts are part of the internal stuff like the TokenScripts and what parts are not)

**Comment by Hanmac** [comment]
@tehdiplomat what is your opinion there?

**Comment by tehdiplomat** [comment]
Yea copying back over the altered map params feels like the right move here.

**Comment by tool4ever** [comment]
@Hanmac 
another case is static abilities that grant abilities aren't correctly changed, e. g. on _Regal Sliver_:
> Sliver creatures you control have "When this creature enters the battlefield, Slivers you control get +1/+1 until end of turn if you're the monarch. Otherwise, you become the monarch."

the affected creatures are correctly changed to the new type but will use the original type for their triggers

should I manually apply the staticAbility host TextChanges inside `getTriggerForStaticAbility`?

**Comment by tool4ever** [comment]
I modified the trigger [here](https://github.com/Card-Forge/forge/blob/548c3d8f9462cbbe5abddf137232f11847540dcd/forge-game/src/main/java/forge/game/staticability/StaticAbilityContinuous.java#L862) like this and it will work:
```
actualTrigger.setIntrinsic(true);
actualTrigger.changeTextIntrinsic(stAb.getChangedTextColors(), stAb.getChangedTextTypes());
actualTrigger.setIntrinsic(false);
```

**Comment by Hanmac** [comment]
> I modified the trigger [here](https://github.com/Card-Forge/forge/blob/548c3d8f9462cbbe5abddf137232f11847540dcd/forge-game/src/main/java/forge/game/staticability/StaticAbilityContinuous.java#L862) like this and it will work:
> 
> ```
> actualTrigger.setIntrinsic(true);
> actualTrigger.changeTextIntrinsic(stAb.getChangedTextColors(), stAb.getChangedTextTypes());
> actualTrigger.setIntrinsic(false);
> ```

setting it to Intrinsic would probably mess up some Text

like would this Mess up "Afterlife" trigger again?

**Comment by tool4ever** [comment]
I don't think so?
they are keyworded and would never get added via `AddTrigger$`

It might be more problematic if I did the same below with `GainsTriggerAbsOf$` for _Idris_
but since she only checks exiled cards we can leave her alone for now?

**Comment by tool4ever** [comment]
The mapParams of the new Trigger are processed via `AbilityUtils.getSVar` (which applies the changes)
but this still leaves the naked SVars for overridingAbilities:
like you quoted above `updateChangedText()` does not modify those

I'll check if I can modify `TriggerHandler.parseTrigger` accordingly but might be too much trouble 🤔

File: forge-game/src/main/java/forge/game/spellability/SpellAbility.java [hunk 1]
```diff
@@ -1166,7 +1169,7 @@ public SpellAbility copy(Card host, Player activ, final boolean lki) {
 
             // don't use setHostCard to not trigger the not copied parts yet
 
-            copyHelper(clone, host);
+            copyHelper(clone, host, lki || keepTextChanges);
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/spellability/SpellAbility.java:1172]: If it's lki we should always keep textChanges imo -
fixes this scenario:
1. alter Human type of _Jirina, Dauntless General_
2. use _Quicksilver Elemental_ on her
3. activate new ability -> it should affect the new type of 1.

File: forge-game/src/main/java/forge/game/card/Card.java [hunk 2]
```diff
@@ -4661,6 +4661,9 @@ public final SpellAbility getSpellAbilityForStaticAbility(final String str, fina
         SpellAbility result = storedSpellAbility.get(stAb, str);
         if (result == null) {
             result = AbilityFactory.getAbility(str, this, stAb);
+            // apply text changes from the statics host
+            result.setIntrinsic(true);
+            result.changeTextIntrinsic(stAb.getChangedTextColors(), stAb.getChangedTextTypes());
```
Comment (Hanmac) [inline forge-game/src/main/java/forge/game/card/Card.java:4666]: what part does need to set it to intrinsic?

i mean can't you just pass a parameter to the changeText function to overwrite it?

Comment (tool4ever) [reply forge-game/src/main/java/forge/game/card/Card.java:4666]: yea I can try that, just wasn't sure how much I would need to pass that down recursively

File: forge-game/src/main/java/forge/game/card/Card.java [hunk 3]
```diff
@@ -4669,9 +4680,20 @@ public final SpellAbility getSpellAbilityForStaticAbility(final String str, fina
     }
 
     public final Trigger getTriggerForStaticAbility(final String str, final StaticAbility stAb) {
+        Map<String, String> changedTextColors = stAb.getChangedTextColors();
+        Map<String, String> changedTextTypes = stAb.getChangedTextTypes();
         Trigger result = storedTrigger.get(stAb, str);
+        if (result != null) {
+            boolean needNewCache = !result.getChangedTextColors().equals(changedTextColors);
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/card/Card.java:4687]: if the Text of the static host gets changed *again* (after it's cached) we might be stuck with the previous version. Is there a better way compared to these checks?

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/Card.java:4687]: move this check to CardTraitsBase

like "sameChangedText(CardTraitBase other)"

Comment (Hanmac) [reply forge-game/src/main/java/forge/game/card/Card.java]: i might have moved to card traits, but this is fine too

**Review by Hanmac — APPROVED** [review]
okay for now i think

**Pull request merged**